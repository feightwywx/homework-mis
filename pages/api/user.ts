export type User = {
  isLoggedIn: boolean
  name: string
  userType: UserType
  token: string
}

export type UserType = 'student' | 'teacher'