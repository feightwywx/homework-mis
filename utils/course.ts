import { createMisConnection } from "./mysql";
import { Course } from "./types";

type CourseRow = {
  id: number;
  name: string;
  time: string;
  ended: number;
  teacherID: number;
  teacherName: string;
}

export async function getStudentCourses(id: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT course.* , teacher.name AS teacherName
    FROM course
    JOIN course_student ON course.id = course_student.studentID 
    JOIN teacher ON course.teacherID = teacher.id
    WHERE course_student.studentID = ?;`,
    [id]
  );
  await conn.end();

  const courses = (rows as Array<CourseRow>).map((row) => {
    return {
      ...row,
      ended: row.ended === 0 ? false : true,
    } as Course;
  });

  return courses;
}

export async function getTeacherCourses(id: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT course.* , teacher.name AS teacherName
    FROM course
    JOIN teacher ON course.teacherID = teacher.id
    WHERE teacher.id = ?;`,
    [id]
  );
  await conn.end();

  const courses = (rows as Array<CourseRow>).map((row) => {
    return {
      ...row,
      ended: row.ended === 0 ? false : true,
    } as Course;
  });

  return courses;
}

export async function getCourseByID(id: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT course.* , teacher.name AS teacherName
    FROM course
    JOIN teacher ON course.teacherID = teacher.id
    WHERE course.id = ?;`,
    [id]
  );
  await conn.end();

  const courses = (rows as Array<CourseRow>).map((row) => {
    return {
      ...row,
      ended: row.ended === 0 ? false : true,
    } as Course;
  });

  return courses[0];
}

export async function getCourses() {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT course.*
    FROM course`
  );
  await conn.end();

  const courses = (rows as Array<CourseRow>).map((row) => {
    return {
      ...row,
      ended: row.ended === 0 ? false : true,
    } as Course;
  });

  return courses;
}
