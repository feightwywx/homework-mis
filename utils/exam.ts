import { ResultSetHeader } from "mysql2";
import { createMisConnection } from "./mysql";
import { Exam, ExamResult } from "./types";

type ExamRow = {
  id: number;
  name: string;
  time: string;
  endtime: string;
  location: string;
  courseID: number;
};

type ExamResultRow = {
  score: number | null;
};
type DetailedExamResultRow = ExamResultRow & {
  id: number;
  studentID: number;
  studentName: string;
};

export async function getExamsByCourseID(coid: number): Promise<Exam[]> {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT *
    FROM exam
    WHERE courseID=?`,
    [coid]
  );
  await conn.end();

  const exams = rows as Array<ExamRow>;

  return exams;
}

export async function getExamByID(eid: number): Promise<Exam> {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT *
    FROM exam
    WHERE id=?`,
    [eid]
  );
  await conn.end();

  const exams = (rows as Array<ExamRow>)[0];

  return exams;
}

export async function getStudentExamScore(
  uid: number,
  eid: number
): Promise<number | null> {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT score
    FROM exam_result
    WHERE studentID=? AND examID=?;`,
    [uid, eid]
  );
  await conn.end();

  const score = (rows as Array<ExamResultRow>)[0].score;
  console.log("score", score);

  return score;
}

export async function getTeacherExamScore(eid: number): Promise<ExamResult[]> {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT exam_result.id, studentID, student.name AS studentName, score
    FROM exam_result
    JOIN student ON student.id = exam_result.studentID
    WHERE exam_result.examID = ?;`,
    [eid]
  );
  await conn.end();

  const score = rows as Array<DetailedExamResultRow>;

  return score;
}

export async function updateExamScore(
  erid: number,
  score: number
): Promise<number> {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `UPDATE exam_result SET score=? WHERE id=?;`,
    [score, erid]
  );
  await conn.end();

  return (rows as ResultSetHeader).affectedRows;
}

export async function insertExamAssign(courseID: number, title: string, location: string, time: string, endtime: string) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'INSERT INTO exam (name, location, time, endtime, courseID) ' +
    'VALUES (?, ?, ?, ?, ?) ',
    [title, location, time, endtime, courseID]
  );
  await conn.end();

  return (rows as ResultSetHeader).insertId
}

export async function insertExamTarget(studentID: number, examID: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'INSERT INTO exam_result (studentID, examID) ' +
    'VALUES (?, ?) ',
    [studentID, examID]
  );
  await conn.end();

  return (rows as ResultSetHeader).affectedRows
}
