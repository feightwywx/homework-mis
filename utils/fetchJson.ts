export default async function fetchJson<JSON = unknown>(
  input: RequestInfo, init?: RequestInit
): Promise<JSON> {
  const res = await fetch(input, init);
  const data = await res.json();

  if (res.ok) return data;

  throw new Error(res.statusText);
}