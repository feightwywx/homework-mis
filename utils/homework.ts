import { ResultSetHeader } from "mysql2";
import { getNowMySqlDateTime, createMisConnection } from "./mysql";
import { StudentHomework, Homework, HomeworkDetailContent, HomeworkStudentDetail, HomeworkTeacherDetailContent } from "./types";

type StudentHomeworkRow = {
  id: number;
  title: string;
  assignment: string;
  time: string;
  deadline: string;
  completed: number;
  name: string;
}

type HomeworkRow = {
  id: number;
  title: string;
  assignment: string;
  time: string;
  deadline: string;
  name: string;
}

type HomeworkDetailRow = {
  title: string;
  assignment: string;
  time: string;
  deadline: string;
}

type HomeworkDetailContentRow = {
  completed: number;
  time: string;
  content: string;
  score: number;
  comment: string;
}

type TeacherHomeworkDetailRow = {
  cid: number;          // ContentId
  name: string;         // 学生名
  time: string;
  completed: number;
  content: string;
  score: number;
  comment: string;
}

export async function getStudentHomeworks(id: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'SELECT homework.id, title, assignment, homework.time, deadline, completed, teacher.name ' +
    'FROM homework JOIN homework_content JOIN teacher ' +
    'ON homework.id=homework_content.id AND homework.teacherID=teacher.id ' +
    'WHERE studentID=?',
    [id]
  )
  await conn.end();
  
  const homeworks = (rows as Array<StudentHomeworkRow>).map((row) => {
    return {
      ...row,
      completed: row.completed === 0 ? false : true,
      teacher: row.name
    } as StudentHomework
  })

  return homeworks
}

export async function getTeacherHomeworks(id: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'SELECT homework.id, title, assignment, homework.time, deadline, teacher.name ' +
    'FROM homework JOIN teacher ' +
    'ON homework.teacherID=teacher.id ' +
    'WHERE teacherID=?',
    [id]
  )
  await conn.end();
  
  const homeworks = (rows as Array<HomeworkRow>).map((row) => {
    return {
      ...row,
      teacher: row.name
    } as Homework
  })

  return homeworks
}

export async function getHomeworkDetail(id: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'SELECT title, assignment, time, deadline ' +
    'FROM homework ' +
    'WHERE id=?',
    [id]
  )
  await conn.end();
  

  return (rows as Array<HomeworkDetailRow>).length !== 0
    ? (rows as Array<HomeworkDetailRow>)[0]
    : null;
}

export async function getStudentHomeworkDetail(hwid: number, stuid: number) {
  const hwDetail = await getHomeworkDetail(hwid)
  if (hwDetail === null) return null;

  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'SELECT content, time, completed, score, comment ' +
    'FROM homework_content ' +
    'WHERE studentID=? AND homeworkID=?',
    [stuid, hwid]
  );
  await conn.end();
  
  if ((rows as Array<HomeworkDetailContentRow>).length === 0) return null;
  const detailContentRow = (rows as Array<HomeworkDetailContentRow>)[0];

  let accomplishment = undefined;
  if (detailContentRow.time && detailContentRow.content) {
    accomplishment = {
      time: detailContentRow.time,
      content: detailContentRow.content
    };
  }

  let judge = undefined;
  if (detailContentRow.score) {
    judge = {
      score: detailContentRow.score,
      comment: detailContentRow.comment
    }
  }

  const result = {
    detail: hwDetail,
    content: {
      completed: detailContentRow.completed === 0 ? false : true,
      accomplishment,
      judge
    } as HomeworkDetailContent
  } as HomeworkStudentDetail;

  return result;
}

export async function updateHomework(hwid: number, stuid: number, content: string) {
  const now = await getNowMySqlDateTime();

  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'UPDATE homework_content ' +
    'SET content=?, time=?, completed=1 ' +
    'WHERE studentID=? AND homeworkID=?',
    [content, now, stuid, hwid]
  );
  await conn.end();

  return (rows as ResultSetHeader).affectedRows
}

export async function getTeacherHomeworkDetail(hwid: number, tid: number) {
  const hwDetail = await getHomeworkDetail(hwid)
  if (hwDetail === null) return null;

  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'SELECT homework_content.id AS cid, student.`name`, homework_content.time, completed, content, score, comment ' +
    'FROM homework_content LEFT JOIN homework ' +
    'ON homework_content.homeworkID=homework.id ' +
    'LEFT JOIN student ON studentID=student.id ' +
    'WHERE teacherID=? AND homeworkID=?',
    [tid, hwid]
  );
  await conn.end();

  if ((rows as Array<TeacherHomeworkDetailRow>).length === 0) return null;
  const detailContentRow = (rows as Array<TeacherHomeworkDetailRow>);

  return {
    detail: hwDetail,
    content: detailContentRow.map((item) => {
      let accomplishment = undefined;
      if (item.time && item.content) {
        accomplishment = {
          time: item.time,
          content: item.content
        };
      }

      let judge = undefined;
      if (item.score || item.comment) {
        judge = {
          score: item.score,
          comment: item.comment
        }
      }

      return {
        contentId: item.cid,
        studentName: item.name,
        completed: item.completed === 0 ? false : true,
        accomplishment,
        judge
      } as HomeworkTeacherDetailContent;
    })
  }
}

export async function setReject(hwid: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'UPDATE homework_content ' +
    'SET completed=0, score=NULL ' +
    'WHERE homeworkID=?',
    [hwid]
  );
  await conn.end();

  return (rows as ResultSetHeader).affectedRows
}

export async function updateJudge(cid: number, score: number, comment: string) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'UPDATE homework_content ' +
    'SET score=?, comment=? ' +
    'WHERE id=?',
    [score, comment, cid]
  );
  await conn.end();

  return (rows as ResultSetHeader).affectedRows
}

export async function insertAssign(id: number, title: string, assignment: string, deadline: string) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'INSERT INTO homework (title, assignment, deadline, teacherID) ' +
    'VALUES (?, ?, ?, ?) ',
    [title, assignment, deadline, id]
  );
  await conn.end();

  return (rows as ResultSetHeader).insertId
}

export async function insertTarget(hwid: number, target: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'INSERT INTO homework_content (completed, studentID, homeworkID) ' +
    'VALUES (0, ?, ?) ',
    [target, hwid]
  );
  await conn.end();

  return (rows as ResultSetHeader).affectedRows
}
