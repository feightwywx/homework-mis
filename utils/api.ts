import { NextApiRequest } from "next";
import { JsonResponse, ResponseResult, UserType } from "./types";
import { getId } from "./user";

export function successResponse<T = ResponseResult>(result?: T) {
  return {
    code: 0, 
    result: result ?? ''
  } as JsonResponse<T>
}

export function failResponse(code: number, result?: ResponseResult) {
  return {
    code, 
    result: result ?? ''
  } as JsonResponse
}

export async function parseIdFromReqest(req: NextApiRequest, userType: UserType) {
  let token = undefined;
  if (req.session.user?.token) {
    token = req.session.user?.token
  } else {
    token = await req.body.token;
  }

  return await getId(userType, token);
}

export const statusCode = {
  ROUTER_PARAM_REQUIRED: 100,
  ROUTER_PARAM_INVALID: 1001,
  BODY_PARAM_REQUIRED: 101,
  URL_PARAM_REQUIRED: 102,
  TOKEN_INVALID: 200,
  NUL_QUERY_DATA: 300,
  NUL_QUERY_PARAM: 301,
  BAD_UPDATE: 400
}
