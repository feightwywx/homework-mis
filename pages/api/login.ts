import type { User } from './user';

import { withIronSessionApiRoute } from 'iron-session/next'
import { sessionOptions } from '../../utils/session'
import { NextApiRequest, NextApiResponse } from 'next'
import { getToken } from '../../utils/token';

export default async function loginRoute(req: NextApiRequest, res: NextApiResponse) {

  const { username, password, usertype } = await req.body

  res.json(await getToken(username, password, usertype));
}