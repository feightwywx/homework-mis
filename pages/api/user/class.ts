import { withIronSessionApiRoute } from "iron-session/next";
import { NextApiRequest, NextApiResponse } from "next";
import { failResponse, parseIdFromReqest, statusCode, successResponse } from "../../../utils/api";
import { sessionOptions } from "../../../utils/session";
import { getAllClasses } from "../../../utils/user";

async function classRoute(req: NextApiRequest, res: NextApiResponse) {
  const id = await parseIdFromReqest(req, 'teacher');
  if (!id) {
    res.json(failResponse(statusCode.TOKEN_INVALID));
    return;
  }

  const result = await getAllClasses()
  if (result) {
    res.json(successResponse(result))
  } else {
    res.json(failResponse(statusCode.NUL_QUERY_DATA))
  }
  return;
}

export default withIronSessionApiRoute(classRoute, sessionOptions);
