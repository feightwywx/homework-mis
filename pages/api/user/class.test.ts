import { createMocks } from 'node-mocks-http';
import type { NextApiRequest, NextApiResponse as OrigNextApiResp } from 'next';
import { classRoute } from './class';
import type { JsonResponse } from '../../../utils/types';
import { initTestDatabase } from '../../../test/utils/initTestDatabase';
import { statusCode } from '../../../utils/api';

interface NextApiResponse<T = any> extends OrigNextApiResp<T> {
  _getJSONData: () => T;
}

describe('/api/user/class endpoint', () => {
  beforeAll(async () => {
    return initTestDatabase()
  })

  const token = '128f3a6727ead7a49aaf2caf4a7f66d29f4cf614ea36c2ce09ade39dfbd3e666af14eb906caf9f5aeddf04bdd52947e0';

  function mockRequestResponse(token?: string) {
    // @ts-expect-error
    const { req, res }: { req: NextApiRequest, res: NextApiResponse<JsonResponse> } = createMocks({
      method: 'GET',
      body: {
        token
      },
    })

    return { req, res }
  }

  it('returns TOKEN_INVALID if !token', async () => {
    const { req, res } = mockRequestResponse(undefined);
    await classRoute(req, res);

    expect(res.statusCode).toBe(200);
    expect(res._getJSONData().code).toEqual(statusCode.TOKEN_INVALID);
  })

  it('if ok', async () => {
    const { req, res } = mockRequestResponse(token);
    await classRoute(req, res);

    expect(res.statusCode).toBe(200);
    expect(res._getJSONData()).toMatchSnapshot();
  })
})