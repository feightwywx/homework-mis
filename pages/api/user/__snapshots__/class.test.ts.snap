// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/user/class endpoint if ok 1`] = `
{
  "code": 0,
  "result": [
    {
      "class": "信息管理与信息系统",
    },
    {
      "class": "测试班级",
    },
  ],
}
`;
